{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2016-08-02T11:18:42.624835",
     "start_time": "2016-08-02T11:18:41.518829"
    },
    "collapsed": false,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [],
   "source": [
    "# %load ~/ipyhead\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "%matplotlib inline\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "We're going to load quantification data from Kallisto. We used the index: `Homo_sapiens.GRCh38.cdna.all.kallisto.idx`\n",
    "\n",
    "Per http://f1000research.com/articles/4-1521/v2 (https://github.com/mikelove/tximport), we should use TPM (not estimated counts, which are sensitive to transcript length) and aggregate to gene level.\n",
    "\n",
    "Here is how to annotate with gene names in Python: https://github.com/hammerlab/cohorts/blob/master/cohorts/load.py#L797-L801\n",
    "\n",
    "We made the index by downloading from http://ftp.ensembl.org/pub/release-79/fasta/homo_sapiens/cdna/Homo_sapiens.GRCh38.cdna.all.fa.gz , which is the same as what's suggested on kallisto website. So use release 79.\n",
    "\n",
    "Though note that the newest release is release 85, from ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/README_ensembl:\n",
    "\n",
    "```\n",
    "#tax_id\tncbi_release\tncbi_assembly\tensembl_release\tensembl_assembly\tdate_compared\n",
    "9606\tHomo sapiens Annotation Release 108\tGRCh38.p7\t85\tGRCh38.p7\t20160801\n",
    "```\n",
    "\n",
    "And http://ftp.ensembl.org/pub/release-85/fasta/homo_sapiens/cdna/ has different files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2016-08-02T11:32:11.617279",
     "start_time": "2016-08-02T11:32:11.360058"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target_id</th>\n",
       "      <th>length</th>\n",
       "      <th>eff_length</th>\n",
       "      <th>est_counts</th>\n",
       "      <th>tpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000415118</td>\n",
       "      <td>8</td>\n",
       "      <td>5.08333</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000448914</td>\n",
       "      <td>13</td>\n",
       "      <td>8.66667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000434970</td>\n",
       "      <td>9</td>\n",
       "      <td>5.69231</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000390577</td>\n",
       "      <td>37</td>\n",
       "      <td>21.37140</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.44565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000437320</td>\n",
       "      <td>19</td>\n",
       "      <td>11.90480</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         target_id  length  eff_length  est_counts      tpm\n",
       "0  ENST00000415118       8     5.08333         0.0  0.00000\n",
       "1  ENST00000448914      13     8.66667         0.0  0.00000\n",
       "2  ENST00000434970       9     5.69231         0.0  0.00000\n",
       "3  ENST00000390577      37    21.37140         1.0  6.44565\n",
       "4  ENST00000437320      19    11.90480         0.0  0.00000"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('output/ERR431617/abundance.tsv', sep='\\t')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2016-08-02T11:23:43.586269",
     "start_time": "2016-08-02T11:23:19.746249"
    },
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting pyensembl\n",
      "  Downloading pyensembl-0.9.5.tar.gz (60kB)\n",
      "\u001b[K    100% |████████████████████████████████| 61kB 1.6MB/s \n",
      "\u001b[?25hCollecting typechecks>=0.0.2 (from pyensembl)\n",
      "  Downloading typechecks-0.0.2.tar.gz\n",
      "Requirement already satisfied (use --upgrade to upgrade): numpy>=1.7 in /home/maxim/anaconda2/lib/python2.7/site-packages (from pyensembl)\n",
      "Requirement already satisfied (use --upgrade to upgrade): pandas>=0.15 in /home/maxim/anaconda2/lib/python2.7/site-packages (from pyensembl)\n",
      "Collecting datacache>=0.4.16 (from pyensembl)\n",
      "  Downloading datacache-0.4.17.tar.gz\n",
      "Collecting memoized-property>=1.0.2 (from pyensembl)\n",
      "  Downloading memoized-property-1.0.2.tar.gz\n",
      "Requirement already satisfied (use --upgrade to upgrade): six>=1.9.0 in /home/maxim/anaconda2/lib/python2.7/site-packages (from pyensembl)\n",
      "Collecting gtfparse>=0.0.3 (from pyensembl)\n",
      "  Downloading gtfparse-0.0.6.tar.gz\n",
      "Collecting serializable (from pyensembl)\n",
      "  Downloading serializable-0.0.6.tar.gz\n",
      "Requirement already satisfied (use --upgrade to upgrade): python-dateutil in /home/maxim/anaconda2/lib/python2.7/site-packages (from pandas>=0.15->pyensembl)\n",
      "Requirement already satisfied (use --upgrade to upgrade): pytz>=2011k in /home/maxim/anaconda2/lib/python2.7/site-packages (from pandas>=0.15->pyensembl)\n",
      "Collecting appdirs>=1.4.0 (from datacache>=0.4.16->pyensembl)\n",
      "  Downloading appdirs-1.4.0-py2.py3-none-any.whl\n",
      "Collecting progressbar33>=2.4 (from datacache>=0.4.16->pyensembl)\n",
      "  Downloading progressbar33-2.4.tar.gz\n",
      "Collecting biopython>=1.65 (from datacache>=0.4.16->pyensembl)\n",
      "  Downloading biopython-1.67.tar.gz (14.3MB)\n",
      "\u001b[K    100% |████████████████████████████████| 14.3MB 74kB/s \n",
      "\u001b[?25hRequirement already satisfied (use --upgrade to upgrade): requests>=2.5.1 in /home/maxim/anaconda2/lib/python2.7/site-packages (from datacache>=0.4.16->pyensembl)\n",
      "Collecting simplejson (from serializable->pyensembl)\n",
      "  Downloading simplejson-3.8.2.tar.gz (76kB)\n",
      "\u001b[K    100% |████████████████████████████████| 81kB 9.5MB/s \n",
      "\u001b[?25hBuilding wheels for collected packages: pyensembl, typechecks, datacache, memoized-property, gtfparse, serializable, progressbar33, biopython, simplejson\n",
      "  Running setup.py bdist_wheel for pyensembl ... \u001b[?25l-\b \b\\\b \bdone\n",
      "\u001b[?25h  Stored in directory: /home/maxim/.cache/pip/wheels/04/69/38/f675315f843ae0d30475fbbdd865622995dc64ae7adee83a3f\n",
      "  Running setup.py bdist_wheel for typechecks ... \u001b[?25l-\b \b\\\b \bdone\n",
      "\u001b[?25h  Stored in directory: /home/maxim/.cache/pip/wheels/f5/9e/66/87ce0b0e4abaf030ccdf2822cdd59cc2b1fffa6584236faf5e\n",
      "  Running setup.py bdist_wheel for datacache ... \u001b[?25l-\b \b\\\b \bdone\n",
      "\u001b[?25h  Stored in directory: /home/maxim/.cache/pip/wheels/f6/37/55/ba59c1f1e2c7df84eca3df207aba1f9ed385bc4d2f674db8d2\n",
      "  Running setup.py bdist_wheel for memoized-property ... \u001b[?25l-\b \bdone\n",
      "\u001b[?25h  Stored in directory: /home/maxim/.cache/pip/wheels/09/42/c6/fa081008bb537f37b06168cb296c94d70f836c49d1f52c78ba\n",
      "  Running setup.py bdist_wheel for gtfparse ... \u001b[?25l-\b \b\\\b \bdone\n",
      "\u001b[?25h  Stored in directory: /home/maxim/.cache/pip/wheels/ed/90/50/007c4aa4ae0e50fffec744aa3357566eb834fae066651737b9\n",
      "  Running setup.py bdist_wheel for serializable ... \u001b[?25l-\b \b\\\b \bdone\n",
      "\u001b[?25h  Stored in directory: /home/maxim/.cache/pip/wheels/24/e1/e2/3e3ee08809d1b6fc3d76584641477be0090b1ca4ddaa150ecf\n",
      "  Running setup.py bdist_wheel for progressbar33 ... \u001b[?25l-\b \bdone\n",
      "\u001b[?25h  Stored in directory: /home/maxim/.cache/pip/wheels/4d/72/a3/a74877f299ef2e9bed86e29ab723bd5adfe7cef6f1d0cfa986\n",
      "  Running setup.py bdist_wheel for biopython ... \u001b[?25l-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \bdone\n",
      "\u001b[?25h  Stored in directory: /home/maxim/.cache/pip/wheels/5a/7f/ff/5e827f85f9647a1cb459b917e4e48044abda908077fc6f8240\n",
      "  Running setup.py bdist_wheel for simplejson ... \u001b[?25l-\b \b\\\b \bdone\n",
      "\u001b[?25h  Stored in directory: /home/maxim/.cache/pip/wheels/e4/32/71/60b361b0d05433eb9d1dd3d47619931c08cc4e387dc494ad3c\n",
      "Successfully built pyensembl typechecks datacache memoized-property gtfparse serializable progressbar33 biopython simplejson\n",
      "Installing collected packages: typechecks, appdirs, progressbar33, biopython, datacache, memoized-property, gtfparse, simplejson, serializable, pyensembl\n",
      "Successfully installed appdirs-1.4.0 biopython-1.67 datacache-0.4.17 gtfparse-0.0.6 memoized-property-1.0.2 progressbar33-2.4 pyensembl-0.9.5 serializable-0.0.6 simplejson-3.8.2 typechecks-0.0.2\n",
      "\u001b[33mYou are using pip version 8.1.1, however version 8.1.2 is available.\n",
      "You should consider upgrading via the 'pip install --upgrade pip' command.\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip install pyensembl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2016-08-02T11:36:53.549840",
     "start_time": "2016-08-02T11:32:37.034383"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-- Running 'install' for EnsemblRelease(release=79, species='homo_sapiens')\n",
      "INFO:root:Fetching /home/maxim/.cache/pyensembl/GRCh38/ensembl79/Homo_sapiens.GRCh38.79.gtf.gz from URL ftp://ftp.ensembl.org/pub/release-79/gtf/homo_sapiens/Homo_sapiens.GRCh38.79.gtf.gz\n",
      "Downloading ftp://ftp.ensembl.org/pub/release-79/gtf/homo_sapiens/Homo_sapiens.GRCh38.79.gtf.gz to /home/maxim/.cache/pyensembl/GRCh38/ensembl79/Homo_sapiens.GRCh38.79.gtf.gz\n",
      "INFO:root:Fetching /home/maxim/.cache/pyensembl/GRCh38/ensembl79/Homo_sapiens.GRCh38.cdna.all.fa.gz from URL ftp://ftp.ensembl.org/pub/release-79/fasta/homo_sapiens/cdna/Homo_sapiens.GRCh38.cdna.all.fa.gz\n",
      "Downloading ftp://ftp.ensembl.org/pub/release-79/fasta/homo_sapiens/cdna/Homo_sapiens.GRCh38.cdna.all.fa.gz to /home/maxim/.cache/pyensembl/GRCh38/ensembl79/Homo_sapiens.GRCh38.cdna.all.fa.gz\n",
      "INFO:root:Fetching /home/maxim/.cache/pyensembl/GRCh38/ensembl79/Homo_sapiens.GRCh38.pep.all.fa.gz from URL ftp://ftp.ensembl.org/pub/release-79/fasta/homo_sapiens/pep/Homo_sapiens.GRCh38.pep.all.fa.gz\n",
      "Downloading ftp://ftp.ensembl.org/pub/release-79/fasta/homo_sapiens/pep/Homo_sapiens.GRCh38.pep.all.fa.gz to /home/maxim/.cache/pyensembl/GRCh38/ensembl79/Homo_sapiens.GRCh38.pep.all.fa.gz\n",
      "Creating database: /home/maxim/.cache/pyensembl/GRCh38/ensembl79/Homo_sapiens.GRCh38.79.gtf.db\n",
      "Reading GTF from /home/maxim/.cache/pyensembl/GRCh38/ensembl79/Homo_sapiens.GRCh38.79.gtf.gz\n",
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_version', 'gene_name', 'gene_source', 'gene_biotype', 'transcript_id', 'transcript_version', 'transcript_name', 'transcript_source', 'transcript_biotype', 'tag', 'transcript_support_level', 'exon_number', 'exon_id', 'exon_version', 'ccds_id', 'protein_id', 'protein_version']\n",
      "Saving DataFrame to /home/maxim/.cache/pyensembl/GRCh38/ensembl79/Homo_sapiens.GRCh38.79.gtf.expanded.csv\n",
      "INFO:root:Creating database /home/maxim/.cache/pyensembl/GRCh38/ensembl79/Homo_sapiens.GRCh38.79.gtf.db containing: gene, Selenocysteine, UTR, exon, stop_codon, CDS, start_codon, transcript\n",
      "INFO:root:Running sqlite query: \"CREATE TABLE gene (seqname TEXT NOT NULL, source TEXT NOT NULL, feature TEXT NOT NULL, start INT NOT NULL, end INT NOT NULL, score FLOAT, strand TEXT NOT NULL, frame TEXT NOT NULL, gene_id TEXT UNIQUE PRIMARY KEY NOT NULL, gene_version TEXT NOT NULL, gene_name TEXT NOT NULL, gene_source TEXT NOT NULL, gene_biotype TEXT NOT NULL, transcript_id TEXT NOT NULL, transcript_version TEXT NOT NULL, transcript_name TEXT NOT NULL, transcript_source TEXT NOT NULL, transcript_biotype TEXT NOT NULL, tag TEXT NOT NULL, transcript_support_level TEXT NOT NULL, exon_number TEXT NOT NULL, exon_id TEXT NOT NULL, exon_version TEXT NOT NULL, ccds_id TEXT NOT NULL, protein_id TEXT NOT NULL, protein_version TEXT NOT NULL)\"\n",
      "INFO:root:Inserting 65217 rows into table gene\n",
      "INFO:root:Creating index on gene (seqname, start, end)\n",
      "INFO:root:Creating index on gene (seqname, start, end, strand)\n",
      "INFO:root:Creating index on gene (gene_name)\n",
      "INFO:root:Creating index on gene (transcript_id)\n",
      "INFO:root:Creating index on gene (transcript_name)\n",
      "INFO:root:Creating index on gene (exon_id)\n",
      "INFO:root:Creating index on gene (protein_id)\n",
      "INFO:root:Creating index on gene (ccds_id)\n",
      "INFO:root:Running sqlite query: \"CREATE TABLE Selenocysteine (seqname TEXT NOT NULL, source TEXT NOT NULL, feature TEXT NOT NULL, start INT NOT NULL, end INT NOT NULL, score FLOAT, strand TEXT NOT NULL, frame TEXT NOT NULL, gene_id TEXT NOT NULL, gene_version TEXT NOT NULL, gene_name TEXT NOT NULL, gene_source TEXT NOT NULL, gene_biotype TEXT NOT NULL, transcript_id TEXT NOT NULL, transcript_version TEXT NOT NULL, transcript_name TEXT NOT NULL, transcript_source TEXT NOT NULL, transcript_biotype TEXT NOT NULL, tag TEXT NOT NULL, transcript_support_level TEXT NOT NULL, exon_number TEXT NOT NULL, exon_id TEXT NOT NULL, exon_version TEXT NOT NULL, ccds_id TEXT NOT NULL, protein_id TEXT NOT NULL, protein_version TEXT NOT NULL)\"\n",
      "INFO:root:Inserting 114 rows into table Selenocysteine\n",
      "INFO:root:Creating index on Selenocysteine (seqname, start, end)\n",
      "INFO:root:Creating index on Selenocysteine (seqname, start, end, strand)\n",
      "INFO:root:Creating index on Selenocysteine (gene_name)\n",
      "INFO:root:Creating index on Selenocysteine (gene_id)\n",
      "INFO:root:Creating index on Selenocysteine (transcript_id)\n",
      "INFO:root:Creating index on Selenocysteine (transcript_name)\n",
      "INFO:root:Creating index on Selenocysteine (exon_id)\n",
      "INFO:root:Creating index on Selenocysteine (protein_id)\n",
      "INFO:root:Creating index on Selenocysteine (ccds_id)\n",
      "INFO:root:Running sqlite query: \"CREATE TABLE UTR (seqname TEXT NOT NULL, source TEXT NOT NULL, feature TEXT NOT NULL, start INT NOT NULL, end INT NOT NULL, score FLOAT, strand TEXT NOT NULL, frame TEXT NOT NULL, gene_id TEXT NOT NULL, gene_version TEXT NOT NULL, gene_name TEXT NOT NULL, gene_source TEXT NOT NULL, gene_biotype TEXT NOT NULL, transcript_id TEXT NOT NULL, transcript_version TEXT NOT NULL, transcript_name TEXT NOT NULL, transcript_source TEXT NOT NULL, transcript_biotype TEXT NOT NULL, tag TEXT NOT NULL, transcript_support_level TEXT NOT NULL, exon_number TEXT NOT NULL, exon_id TEXT NOT NULL, exon_version TEXT NOT NULL, ccds_id TEXT NOT NULL, protein_id TEXT NOT NULL, protein_version TEXT NOT NULL)\"\n",
      "INFO:root:Inserting 290327 rows into table UTR\n",
      "INFO:root:Creating index on UTR (seqname, start, end)\n",
      "INFO:root:Creating index on UTR (seqname, start, end, strand)\n",
      "INFO:root:Creating index on UTR (gene_name)\n",
      "INFO:root:Creating index on UTR (gene_id)\n",
      "INFO:root:Creating index on UTR (transcript_id)\n",
      "INFO:root:Creating index on UTR (transcript_name)\n",
      "INFO:root:Creating index on UTR (exon_id)\n",
      "INFO:root:Creating index on UTR (protein_id)\n",
      "INFO:root:Creating index on UTR (ccds_id)\n",
      "INFO:root:Running sqlite query: \"CREATE TABLE exon (seqname TEXT NOT NULL, source TEXT NOT NULL, feature TEXT NOT NULL, start INT NOT NULL, end INT NOT NULL, score FLOAT, strand TEXT NOT NULL, frame TEXT NOT NULL, gene_id TEXT NOT NULL, gene_version TEXT NOT NULL, gene_name TEXT NOT NULL, gene_source TEXT NOT NULL, gene_biotype TEXT NOT NULL, transcript_id TEXT NOT NULL, transcript_version TEXT NOT NULL, transcript_name TEXT NOT NULL, transcript_source TEXT NOT NULL, transcript_biotype TEXT NOT NULL, tag TEXT NOT NULL, transcript_support_level TEXT NOT NULL, exon_number TEXT NOT NULL, exon_id TEXT NOT NULL, exon_version TEXT NOT NULL, ccds_id TEXT NOT NULL, protein_id TEXT NOT NULL, protein_version TEXT NOT NULL)\"\n",
      "INFO:root:Inserting 1259629 rows into table exon\n",
      "INFO:root:Creating index on exon (seqname, start, end)\n",
      "INFO:root:Creating index on exon (seqname, start, end, strand)\n",
      "INFO:root:Creating index on exon (gene_name)\n",
      "INFO:root:Creating index on exon (gene_id)\n",
      "INFO:root:Creating index on exon (transcript_id)\n",
      "INFO:root:Creating index on exon (transcript_name)\n",
      "INFO:root:Creating index on exon (exon_id)\n",
      "INFO:root:Creating index on exon (protein_id)\n",
      "INFO:root:Creating index on exon (ccds_id)\n",
      "INFO:root:Running sqlite query: \"CREATE TABLE stop_codon (seqname TEXT NOT NULL, source TEXT NOT NULL, feature TEXT NOT NULL, start INT NOT NULL, end INT NOT NULL, score FLOAT, strand TEXT NOT NULL, frame TEXT NOT NULL, gene_id TEXT NOT NULL, gene_version TEXT NOT NULL, gene_name TEXT NOT NULL, gene_source TEXT NOT NULL, gene_biotype TEXT NOT NULL, transcript_id TEXT NOT NULL, transcript_version TEXT NOT NULL, transcript_name TEXT NOT NULL, transcript_source TEXT NOT NULL, transcript_biotype TEXT NOT NULL, tag TEXT NOT NULL, transcript_support_level TEXT NOT NULL, exon_number TEXT NOT NULL, exon_id TEXT NOT NULL, exon_version TEXT NOT NULL, ccds_id TEXT NOT NULL, protein_id TEXT NOT NULL, protein_version TEXT NOT NULL)\"\n",
      "INFO:root:Inserting 69998 rows into table stop_codon\n",
      "INFO:root:Creating index on stop_codon (seqname, start, end)\n",
      "INFO:root:Creating index on stop_codon (seqname, start, end, strand)\n",
      "INFO:root:Creating index on stop_codon (gene_name)\n",
      "INFO:root:Creating index on stop_codon (gene_id)\n",
      "INFO:root:Creating index on stop_codon (transcript_id)\n",
      "INFO:root:Creating index on stop_codon (transcript_name)\n",
      "INFO:root:Creating index on stop_codon (exon_id)\n",
      "INFO:root:Creating index on stop_codon (protein_id)\n",
      "INFO:root:Creating index on stop_codon (ccds_id)\n",
      "INFO:root:Running sqlite query: \"CREATE TABLE CDS (seqname TEXT NOT NULL, source TEXT NOT NULL, feature TEXT NOT NULL, start INT NOT NULL, end INT NOT NULL, score FLOAT, strand TEXT NOT NULL, frame TEXT NOT NULL, gene_id TEXT NOT NULL, gene_version TEXT NOT NULL, gene_name TEXT NOT NULL, gene_source TEXT NOT NULL, gene_biotype TEXT NOT NULL, transcript_id TEXT NOT NULL, transcript_version TEXT NOT NULL, transcript_name TEXT NOT NULL, transcript_source TEXT NOT NULL, transcript_biotype TEXT NOT NULL, tag TEXT NOT NULL, transcript_support_level TEXT NOT NULL, exon_number TEXT NOT NULL, exon_id TEXT NOT NULL, exon_version TEXT NOT NULL, ccds_id TEXT NOT NULL, protein_id TEXT NOT NULL, protein_version TEXT NOT NULL)\"\n",
      "INFO:root:Inserting 751742 rows into table CDS\n",
      "INFO:root:Creating index on CDS (seqname, start, end)\n",
      "INFO:root:Creating index on CDS (seqname, start, end, strand)\n",
      "INFO:root:Creating index on CDS (gene_name)\n",
      "INFO:root:Creating index on CDS (gene_id)\n",
      "INFO:root:Creating index on CDS (transcript_id)\n",
      "INFO:root:Creating index on CDS (transcript_name)\n",
      "INFO:root:Creating index on CDS (exon_id)\n",
      "INFO:root:Creating index on CDS (protein_id)\n",
      "INFO:root:Creating index on CDS (ccds_id)\n",
      "INFO:root:Running sqlite query: \"CREATE TABLE start_codon (seqname TEXT NOT NULL, source TEXT NOT NULL, feature TEXT NOT NULL, start INT NOT NULL, end INT NOT NULL, score FLOAT, strand TEXT NOT NULL, frame TEXT NOT NULL, gene_id TEXT NOT NULL, gene_version TEXT NOT NULL, gene_name TEXT NOT NULL, gene_source TEXT NOT NULL, gene_biotype TEXT NOT NULL, transcript_id TEXT NOT NULL, transcript_version TEXT NOT NULL, transcript_name TEXT NOT NULL, transcript_source TEXT NOT NULL, transcript_biotype TEXT NOT NULL, tag TEXT NOT NULL, transcript_support_level TEXT NOT NULL, exon_number TEXT NOT NULL, exon_id TEXT NOT NULL, exon_version TEXT NOT NULL, ccds_id TEXT NOT NULL, protein_id TEXT NOT NULL, protein_version TEXT NOT NULL)\"\n",
      "INFO:root:Inserting 69881 rows into table start_codon\n",
      "INFO:root:Creating index on start_codon (seqname, start, end)\n",
      "INFO:root:Creating index on start_codon (seqname, start, end, strand)\n",
      "INFO:root:Creating index on start_codon (gene_name)\n",
      "INFO:root:Creating index on start_codon (gene_id)\n",
      "INFO:root:Creating index on start_codon (transcript_id)\n",
      "INFO:root:Creating index on start_codon (transcript_name)\n",
      "INFO:root:Creating index on start_codon (exon_id)\n",
      "INFO:root:Creating index on start_codon (protein_id)\n",
      "INFO:root:Creating index on start_codon (ccds_id)\n",
      "INFO:root:Running sqlite query: \"CREATE TABLE transcript (seqname TEXT NOT NULL, source TEXT NOT NULL, feature TEXT NOT NULL, start INT NOT NULL, end INT NOT NULL, score FLOAT, strand TEXT NOT NULL, frame TEXT NOT NULL, gene_id TEXT NOT NULL, gene_version TEXT NOT NULL, gene_name TEXT NOT NULL, gene_source TEXT NOT NULL, gene_biotype TEXT NOT NULL, transcript_id TEXT UNIQUE PRIMARY KEY NOT NULL, transcript_version TEXT NOT NULL, transcript_name TEXT NOT NULL, transcript_source TEXT NOT NULL, transcript_biotype TEXT NOT NULL, tag TEXT NOT NULL, transcript_support_level TEXT NOT NULL, exon_number TEXT NOT NULL, exon_id TEXT NOT NULL, exon_version TEXT NOT NULL, ccds_id TEXT NOT NULL, protein_id TEXT NOT NULL, protein_version TEXT NOT NULL)\"\n",
      "INFO:root:Inserting 213622 rows into table transcript\n",
      "INFO:root:Creating index on transcript (seqname, start, end)\n",
      "INFO:root:Creating index on transcript (seqname, start, end, strand)\n",
      "INFO:root:Creating index on transcript (gene_name)\n",
      "INFO:root:Creating index on transcript (gene_id)\n",
      "INFO:root:Creating index on transcript (transcript_name)\n",
      "INFO:root:Creating index on transcript (exon_id)\n",
      "INFO:root:Creating index on transcript (protein_id)\n",
      "INFO:root:Creating index on transcript (ccds_id)\n",
      "INFO:root:Running sqlite query: \"CREATE TABLE _datacache_metadata (version INT)\"\n",
      "INFO:root:Running sqlite query: \"INSERT INTO _datacache_metadata VALUES (2)\"\n",
      "INFO:root:Parsing sequences from FASTA file at /home/maxim/.cache/pyensembl/GRCh38/ensembl79/Homo_sapiens.GRCh38.cdna.all.fa.gz\n",
      "INFO:root:Saving sequence dictionary to /home/maxim/.cache/pyensembl/GRCh38/ensembl79/Homo_sapiens.GRCh38.cdna.all.fa.gz.pickle\n",
      "INFO:root:Parsing sequences from FASTA file at /home/maxim/.cache/pyensembl/GRCh38/ensembl79/Homo_sapiens.GRCh38.pep.all.fa.gz\n",
      "INFO:root:Saving sequence dictionary to /home/maxim/.cache/pyensembl/GRCh38/ensembl79/Homo_sapiens.GRCh38.pep.all.fa.gz.pickle\n"
     ]
    }
   ],
   "source": [
    "!pyensembl install --release 79 --species human"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from pyensembl import cached_release"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ensembl_release = cached_release(79)\n",
    "df['sample_id'] = 1\n",
    "df['gene_name'] = df['target_id'].map(lambda t: ensembl_release.gene_name_of_transcript_id(t))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>est_counts</th>\n",
       "      <th>tpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>89.513800</td>\n",
       "      <td>14.958630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>14.071500</td>\n",
       "      <td>0.249757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>A2M</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>16.530498</td>\n",
       "      <td>1.209676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0.076626</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_id gene_name  est_counts        tpm\n",
       "0          1      A1BG   89.513800  14.958630\n",
       "1          1      A1CF   14.071500   0.249757\n",
       "2          1       A2M    0.000000   0.000000\n",
       "3          1     A2ML1   16.530498   1.209676\n",
       "4          1     A2MP1    2.000000   0.076626"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# sum counts, abundance (tpm) across genes\n",
    "simple_summary = df.groupby(['sample_id', 'gene_name'])[['est_counts', 'tpm']].sum().reset_index()\n",
    "simple_summary.to_csv('summary.simple.tsv', sep='\\t', index=None)\n",
    "simple_summary.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Scale the TPM, as in that paper:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 173259 entries, 0 to 173258\n",
      "Data columns (total 7 columns):\n",
      "target_id     173259 non-null object\n",
      "length        173259 non-null int64\n",
      "eff_length    173259 non-null float64\n",
      "est_counts    173259 non-null float64\n",
      "tpm           173259 non-null float64\n",
      "sample_id     173259 non-null int64\n",
      "gene_name     173259 non-null object\n",
      "dtypes: float64(3), int64(2), object(2)\n",
      "memory usage: 9.3+ MB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(False, False)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# eff_length is the right one to use.\n",
    "any(pd.isnull(df.length)), any(pd.isnull(df.eff_length))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How to summarize tpm, from https://github.com/mikelove/tximport/blob/master/R/tximport.R\n",
    "\n",
    "abundanceMat is from tpm. lengthMat is from eff_length. countsMat is from est_counts. These are gene-level.\n",
    "Any mat Tx is transcript-level.\n",
    "\n",
    "```\n",
    "  # summarize abundance and counts\n",
    "  message(\"summarizing abundance\")\n",
    "  abundanceMat <- rowsum(abundanceMatTx, geneId)\n",
    "  message(\"summarizing counts\")\n",
    "  countsMat <- rowsum(countsMatTx, geneId)\n",
    "  message(\"summarizing length\")\n",
    "  \n",
    "  # the next lines calculate a weighted average of transcript length, \n",
    "  # weighting by transcript abundance.\n",
    "  # this can be used as an offset / normalization factor which removes length bias\n",
    "  # for the differential analysis of estimated counts summarized at the gene level.\n",
    "  weightedLength <- rowsum(abundanceMatTx * lengthMatTx, geneId)\n",
    "  lengthMat <- weightedLength / abundanceMat   \n",
    "\n",
    "  # pre-calculate a simple average transcript length\n",
    "  # for the case the abundances are all zero for all samples.\n",
    "  # first, average the tx lengths over samples\n",
    "  aveLengthSamp <- rowMeans(lengthMatTx)\n",
    "  # then simple average of lengths within genes (not weighted by abundance)\n",
    "  aveLengthSampGene <- tapply(aveLengthSamp, geneId, mean)\n",
    "\n",
    "  stopifnot(all(names(aveLengthSampGene) == rownames(lengthMat)))\n",
    "  \n",
    "  # check for NaN and if possible replace these values with geometric mean of other samples.\n",
    "  # (the geometic mean here implies an offset of 0 on the log scale)\n",
    "  # NaN come from samples which have abundance of 0 for all isoforms of a gene, and \n",
    "  # so we cannot calculate the weighted average. our best guess is to use the average\n",
    "  # transcript length from the other samples.\n",
    "  lengthMat <- replaceMissingLength(lengthMat, aveLengthSampGene)\n",
    "  \n",
    "  if (countsFromAbundance != \"no\") {\n",
    "    countsSum <- colSums(countsMat)\n",
    "    if (countsFromAbundance == \"lengthScaledTPM\") {\n",
    "      newCounts <- abundanceMat * rowMeans(lengthMat)\n",
    "    } else {\n",
    "        newCounts <- abundanceMat\n",
    "      }\n",
    "    newSum <- colSums(newCounts)\n",
    "    countsMat <- t(t(newCounts) * (countsSum/newSum))\n",
    "  }\n",
    "    return(list(abundance=abundanceMat, counts=countsMat, length=lengthMat,\n",
    "              countsFromAbundance=countsFromAbundance))\n",
    "\n",
    "\n",
    "# function for replacing missing average transcript length values\n",
    "replaceMissingLength <- function(lengthMat, aveLengthSampGene) {\n",
    "  nanRows <- which(apply(lengthMat, 1, function(row) any(is.nan(row))))\n",
    "  if (length(nanRows) > 0) {\n",
    "    for (i in nanRows) {\n",
    "      if (all(is.nan(lengthMat[i,]))) {\n",
    "        # if all samples have 0 abundances for all tx, use the simple average\n",
    "        lengthMat[i,] <- aveLengthSampGene[i]\n",
    "      } else {\n",
    "          # otherwise use the geometric mean of the lengths from the other samples\n",
    "          idx <- is.nan(lengthMat[i,])\n",
    "          lengthMat[i,idx] <-  exp(mean(log(lengthMat[i,!idx]), na.rm=TRUE))\n",
    "        }\n",
    "    }\n",
    "  }\n",
    "  lengthMat\n",
    "}\n",
    "\n",
    "\n",
    "```\n",
    "\n",
    "Note how the matrices are set up:\n",
    "\n",
    "* rows: transcripts (n)\n",
    "* columns: samples (m)\n",
    "\n",
    "In R:\n",
    "* rowSums returns n*1 -- i.e. sums columns in every row\n",
    "* colSums return 1*m: i.e. sums all rows for every column\n",
    "* similarly, rowMeans returns n*1: takes mean of values in each row.\n",
    "* similarly, colMeans returns m*1: takes mean of values in each column.\n",
    "\n",
    "```\n",
    "> x <- cbind(x1 = 3, x2 = c(4:1, 2:5))\n",
    "> x\n",
    "     x1 x2\n",
    "[1,]  3  4\n",
    "[2,]  3  3\n",
    "[3,]  3  2\n",
    "[4,]  3  1\n",
    "[5,]  3  2\n",
    "[6,]  3  3\n",
    "[7,]  3  4\n",
    "[8,]  3  5\n",
    "> rowSums(x)\n",
    "[1] 7 6 5 4 5 6 7 8\n",
    "> colSums(x)\n",
    "x1 x2 \n",
    "24 24 \n",
    "> rowMeans(x)\n",
    "[1] 3.5 3.0 2.5 2.0 2.5 3.0 3.5 4.0\n",
    "> colMeans(x)\n",
    "x1 x2 \n",
    " 3  3 \n",
    "```\n",
    "\n",
    "but:\n",
    "\n",
    "* rowsum(x, group) means take vertical (column) sums across rows for each value of a grouping variable. returns k (number of groups) * m.\n",
    "\n",
    "```\n",
    "> require(stats)\n",
    "> x <- matrix(runif(100), ncol = 5)\n",
    "> x\n",
    "            [,1]       [,2]        [,3]       [,4]       [,5]\n",
    " [1,] 0.92193476 0.64068964 0.590930078 0.22589474 0.94385005\n",
    " [2,] 0.15434593 0.35520130 0.891485406 0.01536309 0.23128203\n",
    " [3,] 0.65030467 0.56747106 0.583187841 0.50184454 0.25598191\n",
    " [4,] 0.08740973 0.77858838 0.001003955 0.09488344 0.40001838\n",
    " [5,] 0.59932257 0.10921583 0.202659336 0.99391111 0.05304302\n",
    " [6,] 0.37518234 0.15261407 0.201251270 0.39387585 0.06438196\n",
    " [7,] 0.64699723 0.81092956 0.888885059 0.91184237 0.68612899\n",
    " [8,] 0.83465129 0.23701509 0.182450580 0.94085281 0.55303249\n",
    " [9,] 0.67748311 0.39298104 0.743816100 0.04000909 0.04451748\n",
    "[10,] 0.36702365 0.51603955 0.676513819 0.21491027 0.97969625\n",
    "[11,] 0.81330661 0.01567229 0.265442836 0.33373162 0.03803008\n",
    "[12,] 0.42658481 0.73068255 0.072104913 0.64861247 0.49143138\n",
    "[13,] 0.68161538 0.11003679 0.066483887 0.95039580 0.78276828\n",
    "[14,] 0.57063695 0.85563269 0.517796875 0.11461768 0.36473195\n",
    "[15,] 0.89833696 0.53448586 0.053799714 0.35298232 0.71251391\n",
    "[16,] 0.93033111 0.98725333 0.743280945 0.70274502 0.91984371\n",
    "[17,] 0.72468190 0.80205415 0.304485959 0.51483852 0.03668938\n",
    "[18,] 0.29001996 0.60153044 0.976458703 0.72192388 0.14789380\n",
    "[19,] 0.46333744 0.25058092 0.471588748 0.05430856 0.10111760\n",
    "[20,] 0.78955687 0.94806861 0.675796005 0.87945982 0.77918547\n",
    "> group <- sample(1:8, 20, TRUE)\n",
    "> group\n",
    " [1] 5 4 3 5 5 7 8 5 6 4 2 4 5 5 1 4 4 3 3 3\n",
    "> (xsum <- rowsum(x, group))\n",
    "       [,1]       [,2]       [,3]       [,4]       [,5]\n",
    "1 0.8983370 0.53448586 0.05379971 0.35298232 0.71251391\n",
    "2 0.8133066 0.01567229 0.26544284 0.33373162 0.03803008\n",
    "3 2.1932189 2.36765104 2.70703130 2.15753680 1.28417877\n",
    "4 2.6029674 3.39123087 2.68787104 2.09646937 2.65894275\n",
    "5 3.6955707 2.73117843 1.56132471 3.32055558 3.09744417\n",
    "6 0.6774831 0.39298104 0.74381610 0.04000909 0.04451748\n",
    "7 0.3751823 0.15261407 0.20125127 0.39387585 0.06438196\n",
    "8 0.6469972 0.81092956 0.88888506 0.91184237 0.68612899\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# reimplement R, but keep in tidy form: samples are not columns anymore, we have sample ID column instead\n",
    "\n",
    "abundanceMat = df.groupby(['sample_id', 'gene_name'])[['tpm']].sum()\n",
    "df_tmp = df.copy()\n",
    "df_tmp['weightedLength'] = df_tmp['tpm'] * df_tmp['eff_length']\n",
    "weightedLength = df_tmp.groupby(['sample_id', 'gene_name'])[['weightedLength']].sum()\n",
    "lengthMat = (weightedLength['weightedLength']/ abundanceMat['tpm']).reset_index()\n",
    "# make wide again: make the samples be columns. n_genes x n_samples, values are lengths\n",
    "lengthMat = lengthMat.pivot( index='gene_name', columns='sample_id', values=0)\n",
    "\n",
    "## replace missing average transcript length values\n",
    "\n",
    "# first, calculate average transcript length *across samples* for each transcript\n",
    "# but keep gene name in there\n",
    "aveLengthSamp = df.groupby('target_id')[['eff_length', 'gene_name']].agg({\n",
    "        'eff_length': 'mean',\n",
    "        'gene_name': 'first'\n",
    "    })\n",
    "# second, take simple average of lengths within genes, without weighting by abundance\n",
    "# i.e. aggregate those averages by gene now\n",
    "aveLengthSampGene = aveLengthSamp.groupby('gene_name').eff_length.mean()\n",
    "\n",
    "# do the replacement\n",
    "nanRows = lengthMat[lengthMat.isnull().any(axis=1)].index # np.where(pd.isnull(lengthMat))[0] # row indices where null\n",
    "for row_idx in nanRows:\n",
    "    if all(pd.isnull(lengthMat.loc[row_idx])):\n",
    "        # use simple average\n",
    "        lengthMat.loc[row_idx] = aveLengthSampGene.loc[row_idx] # lengthMat.index[row_idx] # look up by gene name\n",
    "    else:\n",
    "        # use geometric mean of lengths from other samples\n",
    "        null_indices = np.where(pd.isnull(lengthMat.loc[row_idx]))[0]\n",
    "        others = np.where(~pd.isnull(lengthMat.loc[row_idx]))[0]\n",
    "        lengthMat.loc[row_idx, null_indices] = np.exp(np.mean(np.log(lengthMat.loc[row_idx, others])))\n",
    "assert not pd.isnull(lengthMat).any().bool()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# length scaled TPM\n",
    "counts_LS = abundanceMat\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(                        weightedLength\n",
       " sample_id gene_name                   \n",
       " 1         A1BG            12330.780428\n",
       "           A1CF             1938.386555\n",
       "           A2M                 0.000000\n",
       "           A2ML1            2277.117735\n",
       "           A2MP1             275.504952\n",
       "           A3GALT2           413.259230\n",
       "           A4GALT            688.761456\n",
       "           A4GNT               0.000000\n",
       "           AAAS             4683.586265\n",
       "           AACS            22349.414328\n",
       "           AACSP1           1229.235312\n",
       "           AADAC               0.000000\n",
       "           AADACL2             0.000000\n",
       "           AADACL3           275.505086\n",
       "           AADACL4           413.256817\n",
       "           AADACP1             0.000000\n",
       "           AADAT             137.752799\n",
       "           AAED1           41436.691593\n",
       "           AAGAB           64055.062744\n",
       "           AAK1            21099.125589\n",
       "           AAMDC            5703.541360\n",
       "           AAMP            47800.183318\n",
       "           AANAT            4008.740330\n",
       "           AAR2            15637.176032\n",
       "           AARD             1931.891835\n",
       "           AARS            21027.261607\n",
       "           AARS2           14123.554899\n",
       "           AARSD1          26668.543935\n",
       "           AARSP1              0.000000\n",
       "           AASDH            9229.434182\n",
       " ...                                ...\n",
       "           ZSWIM7           6247.362860\n",
       "           ZSWIM8         244052.835619\n",
       "           ZUFSP           57993.891116\n",
       "           ZW10             4270.334216\n",
       "           ZWILCH           7670.554086\n",
       "           ZWINT             137.753381\n",
       "           ZXDA            15370.314164\n",
       "           ZXDB            75270.822958\n",
       "           ZXDC            35953.437002\n",
       "           ZYG11A           8997.642869\n",
       "           ZYG11AP1            0.000000\n",
       "           ZYG11B          35897.976431\n",
       "           ZYX             97529.020762\n",
       "           ZZEF1          308009.997360\n",
       "           ZZZ3            38424.735589\n",
       "           bP-21201H5.1      419.842982\n",
       "           bP-21264C1.1      275.505532\n",
       "           bP-21264C1.3        0.000000\n",
       "           bP-2168N6.1         0.000000\n",
       "           bP-2168N6.2         0.000000\n",
       "           bP-2168N6.3         0.000000\n",
       "           bP-2171C21.2        0.000000\n",
       "           bP-2171C21.4        0.000000\n",
       "           bP-2171C21.5      240.202699\n",
       "           bP-2171C21.6        0.000000\n",
       "           bP-2189O9.1       551.010407\n",
       "           bP-2189O9.2       826.518608\n",
       "           bP-2189O9.4       137.752692\n",
       "           pk              11689.719072\n",
       "           yR211F11.2          0.000000\n",
       " \n",
       " [34832 rows x 1 columns],                                tpm\n",
       " sample_id gene_name               \n",
       " 1         A1BG           14.958630\n",
       "           A1CF            0.249757\n",
       "           A2M             0.000000\n",
       "           A2ML1           1.209676\n",
       "           A2MP1           0.076626\n",
       "           A3GALT2         0.409894\n",
       "           A4GALT          0.690401\n",
       "           A4GNT           0.000000\n",
       "           AAAS            4.978180\n",
       "           AACS           21.098401\n",
       "           AACSP1          0.468773\n",
       "           AADAC           0.000000\n",
       "           AADACL2         0.000000\n",
       "           AADACL3         0.079934\n",
       "           AADACL4         0.305293\n",
       "           AADACP1         0.000000\n",
       "           AADAT           0.242643\n",
       "           AAED1          19.328704\n",
       "           AAGAB          35.069570\n",
       "           AAK1            6.134148\n",
       "           AAMDC          18.129237\n",
       "           AAMP           42.022350\n",
       "           AANAT           6.493162\n",
       "           AAR2            7.624400\n",
       "           AARD            1.365747\n",
       "           AARS           16.442283\n",
       "           AARS2           3.555090\n",
       "           AARSD1         51.608180\n",
       "           AARSP1          0.000000\n",
       "           AASDH           3.564563\n",
       " ...                            ...\n",
       "           ZSWIM7          8.230954\n",
       "           ZSWIM8        112.787238\n",
       "           ZUFSP          58.311290\n",
       "           ZW10            1.698065\n",
       "           ZWILCH          3.351467\n",
       "           ZWINT           0.116129\n",
       "           ZXDA            3.723510\n",
       "           ZXDB           13.380100\n",
       "           ZXDC           11.279710\n",
       "           ZYG11A          2.362190\n",
       "           ZYG11AP1        0.000000\n",
       "           ZYG11B          5.740516\n",
       "           ZYX            59.825725\n",
       "           ZZEF1          51.838230\n",
       "           ZZZ3           26.534556\n",
       "           bP-21201H5.1    0.364485\n",
       "           bP-21264C1.1    0.619204\n",
       "           bP-21264C1.3    0.000000\n",
       "           bP-2168N6.1     0.000000\n",
       "           bP-2168N6.2     0.000000\n",
       "           bP-2168N6.3     0.000000\n",
       "           bP-2171C21.2    0.000000\n",
       "           bP-2171C21.4    0.000000\n",
       "           bP-2171C21.5    4.454740\n",
       "           bP-2171C21.6    0.000000\n",
       "           bP-2189O9.1     0.709957\n",
       "           bP-2189O9.2     0.653762\n",
       "           bP-2189O9.4     0.471974\n",
       "           pk              4.611100\n",
       "           yR211F11.2      0.000000\n",
       " \n",
       " [34832 rows x 1 columns])"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# either scaledTPM or lengthScaledTPM\n",
    "\n",
    "def scaledTPM(matrix):\n",
    "    \n",
    "    pass\n",
    "def lengthScaledTPM(matrix):\n",
    "    pass\n",
    "\n",
    "def core(matrix):\n",
    "    pass\n",
    "abundanceMat = df.groupby(['sample_id', 'gene_name'])[['tpm']].sum()#.reset_index()\n",
    "df_tmp = df.copy()\n",
    "df_tmp['weightedLength'] = df_tmp['tpm'] * df_tmp['eff_length']\n",
    "weightedLength = df_tmp.groupby(['sample_id', 'gene_name'])[['weightedLength']].sum()#.reset_index()\n",
    "lengthMat = (weightedLength['weightedLength']/ abundanceMat['tpm']).reset_index()\n",
    "\n",
    "# replace missing average transcript length values\n",
    "# first, calculate average transcript length *across samples*\n",
    "\n",
    "\n",
    "# replace for case where abundances are all 0 for all samples\n",
    "\n",
    "# function for replacing missing average transcript length values\n",
    "def replaceMissingLength(df, aveLengthSampGene):\n",
    "    lengthMat = df.copy()\n",
    "    nanRows = np.where(pd.isnull(lengthMat))[0] # row indices where null\n",
    "    for row_idx in nanRows:\n",
    "        if all(pd.isnull(lengthMat[row_idx])):\n",
    "            # use simple average\n",
    "            lengthMat.loc[row_idx] = aveLengthSampGene[row_idx]\n",
    "        else:\n",
    "            # use geometric mean of lengths from other samples\n",
    "            null_indices = np.where(pd.isnull(lengthMat[row_idx]))\n",
    "            others = np.where(~pd.isnull(lengthMat[row_idx]))\n",
    "            lengthMath.loc[row_idx, null_indices] = np.exp(np.mean(np.log(lengthMat[row_idx, others])))\n",
    "        \n",
    "    return lengthMat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "2",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-26-be8b7c5f8310>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mnp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mwhere\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mpd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0misnull\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mlengthMat\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m2\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m/home/maxim/anaconda2/lib/python2.7/site-packages/pandas/core/frame.pyc\u001b[0m in \u001b[0;36m__getitem__\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   1990\u001b[0m             \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_getitem_multilevel\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1991\u001b[0m         \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1992\u001b[1;33m             \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_getitem_column\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1993\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1994\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m_getitem_column\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m/home/maxim/anaconda2/lib/python2.7/site-packages/pandas/core/frame.pyc\u001b[0m in \u001b[0;36m_getitem_column\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   1997\u001b[0m         \u001b[1;31m# get column\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1998\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mis_unique\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1999\u001b[1;33m             \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_get_item_cache\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   2000\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2001\u001b[0m         \u001b[1;31m# duplicate columns & possible reduce dimensionality\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m/home/maxim/anaconda2/lib/python2.7/site-packages/pandas/core/generic.pyc\u001b[0m in \u001b[0;36m_get_item_cache\u001b[1;34m(self, item)\u001b[0m\n\u001b[0;32m   1343\u001b[0m         \u001b[0mres\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mcache\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1344\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mres\u001b[0m \u001b[1;32mis\u001b[0m \u001b[0mNone\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1345\u001b[1;33m             \u001b[0mvalues\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_data\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1346\u001b[0m             \u001b[0mres\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_box_item_values\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mvalues\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1347\u001b[0m             \u001b[0mcache\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mitem\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mres\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m/home/maxim/anaconda2/lib/python2.7/site-packages/pandas/core/internals.pyc\u001b[0m in \u001b[0;36mget\u001b[1;34m(self, item, fastpath)\u001b[0m\n\u001b[0;32m   3223\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   3224\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[0misnull\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 3225\u001b[1;33m                 \u001b[0mloc\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   3226\u001b[0m             \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   3227\u001b[0m                 \u001b[0mindexer\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0marange\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0misnull\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m/home/maxim/anaconda2/lib/python2.7/site-packages/pandas/indexes/base.pyc\u001b[0m in \u001b[0;36mget_loc\u001b[1;34m(self, key, method, tolerance)\u001b[0m\n\u001b[0;32m   1876\u001b[0m                 \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1877\u001b[0m             \u001b[1;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1878\u001b[1;33m                 \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_maybe_cast_indexer\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1879\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1880\u001b[0m         \u001b[0mindexer\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_indexer\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mmethod\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mmethod\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mtolerance\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mtolerance\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mpandas/index.pyx\u001b[0m in \u001b[0;36mpandas.index.IndexEngine.get_loc (pandas/index.c:4027)\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;32mpandas/index.pyx\u001b[0m in \u001b[0;36mpandas.index.IndexEngine.get_loc (pandas/index.c:3891)\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;32mpandas/hashtable.pyx\u001b[0m in \u001b[0;36mpandas.hashtable.PyObjectHashTable.get_item (pandas/hashtable.c:12408)\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;32mpandas/hashtable.pyx\u001b[0m in \u001b[0;36mpandas.hashtable.PyObjectHashTable.get_item (pandas/hashtable.c:12359)\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;31mKeyError\u001b[0m: 2"
     ]
    }
   ],
   "source": [
    "np.where(pd.isnull(lengthMat[2]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  },
  "nav_menu": {},
  "toc": {
   "navigate_menu": true,
   "number_sections": true,
   "sideBar": true,
   "threshold": 6,
   "toc_cell": false,
   "toc_section_display": "block",
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
